Strip alpha from inner skin parts

Jerozgen opened the pull request:
Before:
<img width="1174" height="1030" alt="before" src="https://github.com/user-attachments/assets/a4f5bb4c-cd5f-45a3-b586-1a447b5001b9" />
After:
<img width="1174" height="1030" alt="after" src="https://github.com/user-attachments/assets/63befd4f-3515-4a59-95a2-bae58d8cfd85" />

AlexTMjugador commented on packages/app-lib/src/api/minecraft_skins/png_util.rs:
```diff
@@ -151,6 +151,11 @@ pub async fn normalize_skin_texture(
                 png_reader.info(),
             )?;
         }
+        strip_alpha_from_inner_parts(
+            &mut texture_buf,
+            texture_buf_color_type,
+            png_reader.info(),
+        );
```
Supplementing the current skin texture normalization process with this function is a solid start, but a few adjustments are still needed:

- I find the name `strip_alpha_from_inner_parts` to be misleading, as after the function returns, the alpha channel is still present in the texture buffer: it's just set to 255 (i.e., fully opaque) for every pixel in the affected regions. A clearer name would be `make_inner_parts_opaque`, which better reflects this behavior and aligns more closely with Mojang's `setNoAlpha` name for the corresponding method in the game client.
- For legacy-format skins, the game performs an additional normalization step which Mojang referred to as the "Notch transparency hack" (from Mojang's `doNotchTransparencyHack` method name). If we are to precisely match vanilla client skin transparency handling, we should also include this step, which I often found was not done by third-party skin manipulation programs while I originally developed this feature.
  - This step checks all pixels within the square defined by the opposite corners of (32, 0) and (64, 32). Then, if every pixel has an alpha value greater than or equal to 128, all of them are made transparent (alpha set to 0). For more details, check the implementation of the `doNotchTransparencyHack` in any modern Minecraft client version.
  - Crucially, this step must be applied before making inner parts opaque (the step you implemented). Otherwise, the results can differ from those of the vanilla client. (The vanilla client interleaves these two steps, making the first inner part opaque, executing the Notch transparency hack, and then making the remaining inner parts opaque, but since the first inner part doesn't overlap with that of the Notch transparency hack, it’s cleaner in our codebase to perform the hack first.)

AlexTMjugador reviewed (changes requested):
For the Modrinth App, I think it makes the most sense to render skins as closely as possible to how the vanilla client does, which is a baseline most users will find reasonable. Since mods can change many aspects of skin rendering in countless ways, trying to account for those variations sounds both impractical and outside our scope.

When I first wrote this code, I chose not to implement transparency handling because the time required to fully understand Mojang's code and test the edge cases related to transparency didn't seem worth it, especially given that many third-party tools get it wrong anyway (so user expectations weren't clear to me, as @IMB11 noted, albeit for different reasons). However, the fact that someone submitted a PR to adjust the behavior to match the vanilla client is a strong signal that we should handle it that way :smile:

AlexTMjugador commented on packages/app-lib/src/api/minecraft_skins/png_util.rs:
```diff
@@ -283,6 +314,28 @@ fn copy_rect_mirror_horizontally<PixelType: NoUninit + AnyBitPattern>(
+    x2: usize,
+    y2: usize,
+) {
+    if texture_color_type.samples() != 4 {
+        return;
+    }
```
At this point, a valid skin texture buffer can be in one of the following formats: Gray (1 channel), Gray + Alpha (2 channels), RGB (3 channels), or RGBA (4 channels). Therefore, this function should handle the Gray + Alpha case in addition to RGBA.

Jerozgen commented:
A skin texture buffer is now converted to RGBA. This simplifies other code and allows "Notch transparent hack" to add transparency even if the original skin image is RGB or Grayscale. Also I've added 2 more skins to the `normalize_skin_texture_works` test

Jerozgen commented:
Done!
